transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java

```diff
@@ -0,0 +1,217 @@
+    private static final Type HANDLER_METADATA_LIST_TYPE = new TypeToken<List<HandlerMetadata>>() {
+    }.getType();
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    private ChannelHandlerMetadataUtil() {
+    }
+
+    public static void generateMetadata(String projectRelativeResourcePath, String... packageNames) {
+        Set<Class<? extends ChannelHandler>> subtypes = findChannelHandlerSubclasses(packageNames);
+
+        if (Arrays.asList(packageNames).contains("io.netty5.channel")) {
+            // We want the metadata for the ChannelHandler itself too
+            subtypes.add(ChannelHandler.class);
+        }
+
+        Set<HandlerMetadata> handlerMetadata = subtypes.stream()
+                .map(type -> new HandlerMetadata(type.getName(), new Condition(type.getName()), true))
+                .collect(Collectors.toSet());
+
+        Path existingMetadataPath = new File(projectRelativeResourcePath).toPath().toAbsolutePath();
+        if (!Files.exists(existingMetadataPath)) {
+            if (handlerMetadata.size() == 0) {
+                return;
+            }
+
+            System.err.println("Native Image reflection metadata is required for handlers in this project. " +
+                    "This metadata was not found under " + existingMetadataPath);
+            System.err.println("Please create this file with the following content: ");
+            System.err.println(getMetadataJsonString(handlerMetadata));
+            Assertions.fail();
+        }
+
+        String existingMetadataJson = "";
+        try {
+            existingMetadataJson = Files.readString(existingMetadataPath);
+        } catch (IOException e) {
+            Assertions.fail("Failed to open the native-image metadata file at: " + existingMetadataPath, e);
+        }
+        List<HandlerMetadata> existingMetadata = gson.fromJson(existingMetadataJson, HANDLER_METADATA_LIST_TYPE);
+
+        Set<HandlerMetadata> newMetadata = new HashSet<>(handlerMetadata);
+        newMetadata.removeAll(existingMetadata);
+
+        Set<HandlerMetadata> removedMetadata = new HashSet<>(existingMetadata);
+        removedMetadata.removeAll(handlerMetadata);
+
+        if (!newMetadata.isEmpty() || !removedMetadata.isEmpty()) {
+            System.err.println("In the native-image handler metadata file at " + existingMetadataPath);
+
+            if (!newMetadata.isEmpty()) {
+                System.err.println("The following new metadata must be added:\n");
+                System.err.println(getMetadataJsonString(newMetadata));
+                System.err.println();
+            }
+            if (!removedMetadata.isEmpty()) {
+                System.err.println("The following metadata must be removed:\n");
+                System.err.println(getMetadataJsonString(removedMetadata));
+                System.err.println();
+            }
+
+            System.err.println("Expected metadata file contents:\n");
```

Comment:
Maybe we could automatically update the metadata file here, rather than printing a message and letting the users do it

Reply:
I think its fine for now